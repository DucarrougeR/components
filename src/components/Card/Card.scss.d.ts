export const card: string
export const isRaised: string
export const isStacked: string
