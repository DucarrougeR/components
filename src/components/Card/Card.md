
<div class="doc-section-divider"></div>

### A basic Card

At its most basic, a `Card` is a slightly styled container that organizes groups of content, usually linking to a more in-depth view of that content. It can have elevation and a border. A default `Card` makes no assumptions about the content you place inside of it or the size of your card, but you should follow the guidelines to be sure you use them effectively.


```js

// A Card with Block component to add spacing
// Try typing some content into the Block or Card to see how it handles content by default

<Card raised>
  <Block p="4xl"></Block>
</Card>
```

<div class="doc-section-divider"></div>

### A Card with content

Using the [`CardContent`](/#!/CardContent) component quickly gives your content consistent spacing inside your `Card`. You can then use other components to layout your `Card`.

```js
<Card raised>
  <CardContent>
    <Heading size="2">ðŸŽ‰ Looker Release Notes ðŸŽ‰</Heading>
    <Heading level="4" size="5">Read all about our latest features</Heading>
    <Text size="6" mode="subdued">Last updated 3 days ago</Text>
  </CardContent>
</Card>
```

<div class="doc-section-divider"></div>

### A group of media Cards

A few common patterns for `Cards` are displaying them in groups and using images to reinforce the content of the card.

```js
<div style={{display: 'flex'}}>
  <div style={{flex: 1, margin: '10px'}}>
    <Card raised>
      <CardMedia image="https://placeimg.com/640/480/nature" title="Summer Nature"></CardMedia>
      <CardContent>
        <Text size="6" transform="upper" weight="semi-bold" mode="subdued">Summer</Text>
        <Heading level="4" size="3" weight="semi-bold" truncate>Life in The Great Outdoors</Heading>
        <Block mt="xs">
          <Text size="5">10 reasons to get off the couch and head outside this summer.</Text>
        </Block>
      </CardContent>
    </Card>
  </div>
  <div style={{flex: 1, margin: '10px'}}>
    <Card raised>
      <CardMedia image="https://placeimg.com/630/480/nature" title="A Scenic Valley"></CardMedia>
      <CardContent>
        <Text size="6" transform="upper" weight="semi-bold" mode="subdued">Explore</Text>
        <Heading level="4" size="3" weight="semi-bold" truncate>Best Scenic Hikes</Heading>
        <Block mt="xs">
          <Text size="5">Looking for a new place to trailblaze? Make sure it has a great view!</Text>
        </Block>
      </CardContent>
    </Card>
  </div>
  <div style={{flex: 1, margin: '10px'}}>
    <Card raised>
      <CardMedia image="https://placeimg.com/620/480/nature" title="Relaxing Views"></CardMedia>
      <CardContent>
        <Text size="6" transform="upper" weight="semi-bold" mode="subdued">Relax</Text>
        <Heading level="4" size="3" weight="semi-bold" truncate>Mindfull Wilderness</Heading>
        <Block mt="xs">
          <Text size="5">Find a place to find your self.</Text>
        </Block>
      </CardContent>
    </Card>
  </div>
</div>
```

