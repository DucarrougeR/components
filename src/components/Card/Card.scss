.card {
  display: flex;
  flex-direction: column;
  background: #fff;
  border-radius: 4px;
  overflow: hidden;
  min-width: 200px;
  border: 1px solid #E4E5E6
}

.is-raised {
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.07);
}

.is-stacked {
  overflow: hidden;
}
