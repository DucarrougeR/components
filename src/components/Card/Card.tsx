import * as React from 'react'
import { styled, withTheme } from '../../style'
import { ThemedProps } from '../../types'
import { Box, BoxProps } from '../Box'

export interface CardProps extends BoxProps<HTMLDivElement> {
  raised?: boolean
}

const InternalCard: React.FC<ThemedProps<CardProps>> = ({
  raised,
  ...props
}) => (
  <Box
    bg="palette.white"
    borderRadius={props.theme.radii.medium}
    border={`solid 1px ${
      props.theme.colors.semanticColors.primary.borderColor
    }`}
    display="flex"
    flexDirection="column"
    height="100%"
    min-width="200px"
    overflow="hidden"
    {...props}
  >
    {props.children}
  </Box>
)

const InternalCardThemed = withTheme(InternalCard)

const CardFactory = React.forwardRef((props: CardProps, ref) => (
  <InternalCardThemed
    innerRef={ref as React.RefObject<HTMLElement>}
    {...props}
  />
))

// prettier-ignore
export const Card = styled<CardProps>(CardFactory)`
  box-shadow: ${props => (props.raised ? props.theme.shadows[1] : 'none')};
  transition: border ${props => props.theme.transitions.durationQuick}
    ${props => props.theme.easings.ease},
    box-shadow ${props => props.theme.transitions.durationQuick}
    ${props => props.theme.easings.ease};
  &:hover {
    border-color: ${props => props.theme.colors.palette.charcoal300};
    box-shadow: ${props => (props.raised ? props.theme.shadows[2] : 'none')};
  }
`
