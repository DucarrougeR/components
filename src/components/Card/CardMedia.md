```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/PUkKDfseVgoPGNJng7TJ2TIJ/Card"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Card/CardMedia.tsx"
  feedbackTitle="Card Media Component Feedback" />
```

### Displaying an image in your card

A common pattern for `Cards` is to display an image that reinforces what the content is about.

The `CardMedia` component accepts and `image` property that will render the image as a background image and the optional `title` property used to give the image an accessible title if needed.

```js
<Box>
  <Card>
    <CardMedia image="http://fillmurray.com/200/300" title="Bill 1"></CardMedia>
    <CardContent>
      <Heading>A wild Bill appears</Heading>
    </CardContent>
  </Card>
</Box>
<Box mt="medium">
  <Card>
    <CardMedia image="http://fillmurray.com/400/300" title="Bill 2"></CardMedia>
    <CardContent>
      <Heading>Another Bill spoting</Heading>
    </CardContent>
  </Card>
</Box>
```
