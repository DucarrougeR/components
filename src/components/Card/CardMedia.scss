.card-media {
  display: block;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
  overflow: hidden;
  // 16:9 aspect ratio
  height: 0;
  padding-top: 56%;
}
