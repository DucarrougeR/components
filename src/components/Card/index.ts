export { Card } from './Card'
export { CardContent } from './CardContent'
export { CardMedia } from './CardMedia'
