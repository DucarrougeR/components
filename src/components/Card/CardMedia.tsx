import * as React from 'react'
const classNames = require('classnames')
import * as styles from './CardMedia.scss'

export interface CardMediaProps {
  image: string
  title?: string
  className?: string | undefined
}

interface PropertyBag {
  [key: string]: any
}

/** Displays an image that reinforces the Card's content */
export const CardMedia: React.SFC<CardMediaProps> = ({ className, image, ...args }) => {

  const styleableProps: PropertyBag = {

  }

  const cardImage = { backgroundImage: `url(${image})` }

  return (
    <div className={classNames(styles.cardMedia, className, styleableProps)} style={cardImage} {...args} />
  )
}
