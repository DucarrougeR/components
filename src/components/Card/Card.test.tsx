import * as React from 'react'
import { Card } from './Card'
import { CardMedia } from './CardMedia'
import { create } from 'react-test-renderer'
import 'jest-styled-components'
import theme from '../../themes'

test('A Card', () => {
  const component = create(<Card theme={theme}>ðŸ¥‘</Card>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A raised Card', () => {
  const component = create(
    <Card raised={true} theme={theme}>
      ðŸ¥‘
    </Card>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A raised Card with a CardMedia block', () => {
  const component = create(
    <Card raised={true} theme={theme}>
      <CardMedia image="http://placekitten.com/200/300" />
    </Card>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
