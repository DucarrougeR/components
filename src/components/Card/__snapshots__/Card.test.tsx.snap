// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Card 1`] = `
<div
  className=""
>
  ðŸ¥‘
</div>
`;

exports[`A raised Card 1`] = `
<div
  className="undefined"
>
  ðŸ¥‘
</div>
`;

exports[`A raised Card with a CardMedia block 1`] = `
<div
  className="undefined"
>
  <div
    className=""
    style={
      Object {
        "backgroundImage": "url(http://placekitten.com/200/300)",
      }
    }
  />
</div>
`;
